lib/SILGen/SILGenLValue.cpp

```diff
@@ -4605,6 +4674,13 @@ void SILGenFunction::emitAssignLValueToLValue(SILLocation loc, LValue &&src,
     return;
   }
 
+  // If this is an assignment to a distributed actor's actorSystem, push
+  // a clean-up to also initialize its id.
+  // FIXME: in what situations does this lvalue to lvalue assign happen?
+  if (LLVM_UNLIKELY(DistActorCtorContext) &&
+      referencesVar(**(dest.end()-1), DistActorCtorContext->actorSystem))
+    Cleanups.pushCleanup<DistActorIDInitializer>(*DistActorCtorContext);
```

Comment:
Is this via cleanup actually right...? I thought cleanups are done at the end of a scope?

(Just trying to understand this better)

Reply:
Cleanups should be emitted in-order at the end of the evaluation scope in which they were pushed. The caller of `emitAssignToLValue` sets one up for the evaluation of the assignment, but there is also an ArgumentScope established within this function. I might move this clean-up to be just before that ArgumentScope, but otherwise I think it's OK to conditionally push a clean-up here.

More generally, I went with the clean-up approach (for now) because I wanted to make sure we property handle situations like these:

```
init() throws {
  self.actorSystem = try getActorSystem()
}

init() {
  do {
    self.actorSystem = try getActorSystem()
  } catch {
    self.actorSystem = defaultSystem
  }
}
```

Since clean-ups can tell you whether it's being emitted on the unwind side or not. On the unwind / error path of the `try_apply` of `getActorSystem`, we should _not_ try to assign an identity, since we have no actor system (something new that was previously guaranteed with the old approach!). This _might_ mean we need to tweak how `actorReady` calls get injected for synchronous inits (to avoid it on the unwind path), but I'm not sure yet.

Reply:
Missed to thank here; but thanked offline -- thanks for the explanation, makes sense!